main.rb

```diff
@@ -0,0 +1,122 @@
+# Enumerable
+module Enumerable
+  # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+
+  def my_each
+    return to_enum(__method__) unless block_given?
+
+    if is_a?(Array)
+      size.times do |i|
+        yield self[i]
+      end
+    else size.times do |i|
+      yield [keys[i], values[i]]
+    end
+    end
+  end
+
+  def my_each_with_index
+    return to_enum(__method__) unless block_given?
+
+    size.times do |i|
+      yield self[i], i
+    end
+  end
+
+  def my_select
+    return to_enum(__method__) unless block_given?
+
+    array = []
+    my_each do |i|
+      array << i if yield i
+    end
+    array
+  end
+
+  def my_all?(*args)
+    if block_given?
+      my_each { |i| return false unless yield i }
+    elsif args != [] && args[0].class != Regexp
+      my_each { |i| return false unless i.is_a?(args[0]) }
+    elsif args[0].class == Regexp
+      my_each { |i| return false if (args[0] =~ i).nil? }
+    else
+      my_each { |i| return false unless i.nil? }
+    end
+    true
+  end
```

Comment:
- [ ] should have working `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

-  returns false if the block returns false or nil::range

```ruby
#given
false_block = proc { |num| num > 100 }
(0..10).my_all?(&false_block) == (0..10).all?(&false_block)
```
Yours is throwing an error
![image](https://user-images.githubusercontent.com/52209067/85828929-a90cbf00-b746-11ea-9902-e2eb39618e77.png)

Real all?
![image](https://user-images.githubusercontent.com/52209067/85828951-b629ae00-b746-11ea-8163-166bd0086a96.png)

- when no block or argument is given returns true when none of the collection members are false or nil
```ruby
[1, 2, 3, 5].my_all? == [1, 2, 3, 5].all?
```
Yours is giving the wrong answer
![image](https://user-images.githubusercontent.com/52209067/85829521-05bca980-b748-11ea-8a17-4dfebacd5466.png)

Real all?
![image](https://user-images.githubusercontent.com/52209067/85829561-1b31d380-b748-11ea-9859-1d1b12a844b5.png)

- when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
```ruby
[1, 2, 3, 5].my_all?(3) == [1, 2, 3, 5].all?(3)
```
Yours is giving the wrong answer
![image](https://user-images.githubusercontent.com/52209067/85829964-fab64900-b748-11ea-8903-392a1b509b28.png)

Real all?
![image](https://user-images.githubusercontent.com/52209067/85829978-0570de00-b749-11ea-866c-a01fbd39a353.png)